pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -730,89 +732,127 @@ func (service *SecurityPolicyService) deleteVPCSecurityPolicy(sp types.UID, isGC
 	securityPolicyStore, ruleStore, groupStore := service.getSecurityPolicyResourceStores()
 	infraGroupStore, infraShareStore, projectGroupStore, projectShareStore := service.getVPCShareResourceStores()
 
-	// For normal SecurityPolicy deletion process, which means that SecurityPolicy has corresponding NSX SecurityPolicy object
+	// For normal SecurityPolicy deletion process, which means that SecurityPolicy has a corresponding NSX SecurityPolicy object
 	// And for SecurityPolicy GC or cleanup process, which means that SecurityPolicy doesn't exist in K8s any more
-	// but still has corresponding NSX SecurityPolicy object.
+	// but still has a corresponding NSX SecurityPolicy object.
 	// We use SecurityPolicy's UID from store to get NSX SecurityPolicy object
 	indexScope := common.TagValueScopeSecurityPolicyUID
 	if createdFor == common.ResourceTypeNetworkPolicy {
 		indexScope = common.TagScopeNetworkPolicyUID
 	}
-	existingSecurityPolices := securityPolicyStore.GetByIndex(indexScope, string(sp))
-	existingNsxInfraShares := infraShareStore.GetByIndex(indexScope, string(sp))
-	existingNsxInfraShareGroups := infraGroupStore.GetByIndex(indexScope, string(sp))
+	existingSecurityPolices := securityPolicyStore.GetByIndex(indexScope, string(spUID))
+	existingNsxInfraShares := infraShareStore.GetByIndex(indexScope, string(spUID))
+	existingNsxInfraShareGroups := infraGroupStore.GetByIndex(indexScope, string(spUID))
+	// There is no NSX groups/rules in the security policy retrieved from securityPolicy store.
+	// The groups/rules associated the deleting security policy can only be gotten from group/rule store.
+	existingGroups := groupStore.GetByIndex(indexScope, string(spUID))
+	service.markDeleteGroups(existingGroups, nsxGroups, spUID)
+
```

Comment:
This function `deleteVPCSecurityPolicy` is too long. In the next refactor patch, please consider to move L743-L749, L761, L771-L775, L777-L786, L801-L809 out and assemble them to a new function `markSecurityPolicyResourcesDelete` and its returned values can be existingXXXs and vpcInfo.